keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java

```diff
@@ -158,37 +163,24 @@ public PartialDerivatives subtract(PartialDerivatives subtraction, long[] ofShap
         return new PartialDerivatives(subtracted);
     }
 
-    private static Map<VertexId, DoubleTensor> cloneWithCorrectShape(Map<VertexId, DoubleTensor> infinitesimals, long[] ofShape) {
+    private static Map<VertexId, DoubleTensor> cloneWithCorrectShape(Map<VertexId, DoubleTensor> infinitesimals,
+                                                                     boolean ofIsScalar,
+                                                                     long[] resultShape) {
 
         Map<VertexId, DoubleTensor> clone = new HashMap<>();
         for (Map.Entry<VertexId, DoubleTensor> entry : infinitesimals.entrySet()) {
             VertexId k = entry.getKey();
             DoubleTensor v = entry.getValue();
 
-            if (ofShape == null || ofShapeMatches(ofShape, v.getShape())) {
-                clone.put(k, v);
+            if (ofIsScalar) {
+                clone.put(k, DoubleTensor.zeros(TensorShape.concat(resultShape, v.getShape())).plus(v));
             } else {
-                clone.put(k, DoubleTensor.zeros(shapeWrtScalar(ofShape, v.getShape())).plus(v));
+                clone.put(k, v);
             }
         }
```

Comment:
I would find this clearer, with just one call to `clone.put`:

            VertexId k = entry.getKey();
            DoubleTensor v = entry.getValue();

            if (ofIsScalar) {
                v = DoubleTensor.zeros(TensorShape.concat(resultShape, v.getShape())).plus(v);
            }

            clone.put(k, v);

You could even use _ternary if_ so that `v` can be final, but that's a matter of taste.

Reply:
I'd like ternary if both branches were simple. The concat on one side makes the ternary too complex. I try to make local variables final is the only reason it's the way it is. I think the complexity is similar for the `v` reassignment and for the put for each branch. I've gone with the reassignment because I don't mind either way.